test/e2e_node/node_pid_limit_test.go

```diff
@@ -0,0 +1,179 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e_node
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/pkg/features"
+	kubeletconfig "k8s.io/kubernetes/pkg/kubelet/apis/config"
+	"k8s.io/kubernetes/pkg/kubelet/stats/pidlimit"
+	"k8s.io/kubernetes/test/e2e/framework"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+)
+
+const (
+	podCompletionTimeout = 5 * time.Minute
+	minPidsNeededToTest  = 1024
+	desiredPIDSSmall     = 100
+)
+
+func setupNodePidLimitTest(desiredMaxPIDS int) func(initialConfig *kubeletconfig.KubeletConfiguration) {
+	pidLimits, err := pidlimit.Stats()
+	maxPIDS := 0
+	if err == nil && pidLimits != nil && pidLimits.MaxPID != nil {
+		maxPIDS = int(*pidLimits.MaxPID)
+	}
+	return func(initialConfig *kubeletconfig.KubeletConfiguration) {
+		if maxPIDS < minPidsNeededToTest {
+			framework.Skipf("Too few PIDs available on system for test to run")
+		}
+		systemReservedPIDS := maxPIDS - desiredMaxPIDS - (desiredMaxPIDS * 2)
+		kubeReservedPIDS := desiredMaxPIDS * 2
+		initialConfig.FeatureGates[string(features.SupportNodePidsLimit)] = true
+		if initialConfig.SystemReserved == nil {
+			initialConfig.SystemReserved = map[string]string{}
+		}
+		initialConfig.SystemReserved[string(pidlimit.PIDs)] = strconv.Itoa(systemReservedPIDS)
+		if initialConfig.KubeReserved == nil {
+			initialConfig.KubeReserved = map[string]string{}
+		}
+		initialConfig.KubeReserved[string(pidlimit.PIDs)] = strconv.Itoa(kubeReservedPIDS)
+	}
+}
+
+var _ = framework.KubeDescribe("SingleBigPodExceedLimit [Slow] [Serial] [Disruptive] [Feature:NodePidsLimit][NodeFeature:NodePidsLimit]", func() {
+	f := framework.NewDefaultFramework("pid-exceed-kube-limit")
+	Context(fmt.Sprintf("with limit of %d available PIDs", desiredPIDSSmall), func() {
+		tempSetCurrentKubeletConfig(f, setupNodePidLimitTest(desiredPIDSSmall))
+		runNodePidLimitTest(f, true, true, pidConsumingPods("single-big-pod-fail", 1, 150, 60))
```

Comment:
what do we expect to change between these test-cases?  Usually we just run a single test unless we are testing different behaviors with each.  Am I missing some dimension we care about?

Reply:
I suppose I could just use two: one case with multiple pods that sum to more than the limit (to distinguish between per-pod limiting and per-node limiting) and one with a single pod of the same size, under the limit (as a negative test case).

Personally I like to try to bracket the test space to some degree.  Obviously that can become unmanageable beyond a certain point, and I certainly understand if we need to keep the e2e's within bounds.

Reply:
Let me know if you want me to change this, or if you're OK with my leaving it as is.

Reply:
@dashpole also please let me know when you're done with the review so I can collapse out most of the change sets.

Reply:
I would prefer just testing the two variations.  You should also add a comment for each to explain what it is testing.

Feel free to squash.  I have just been looking at the whole set of changes anyways

Reply:
@dashpole done